golem-rib/src/parser/call.rs

```diff
@@ -0,0 +1,1030 @@
+use combine::error::Commit;
+use combine::parser::char::{alpha_num, string};
+use combine::parser::repeat::take_until;
+use combine::stream::easy;
+use combine::{
+    parser::char::{char, spaces},
+    ParseError,
+};
+
+use combine::{any, attempt, between, choice, many1, optional, parser, token, Parser};
+
+use crate::expr::Expr;
+use crate::function_name::{
+    ParsedFunctionName, ParsedFunctionReference, ParsedFunctionSite, SemVer,
+};
+use crate::parser::rib_expr::rib_expr;
+use combine::sep_by;
+
+// A call can be a function or constructing an anonymous variant at the type of writing Rib which user expects to work at runtime
+pub fn call<'t>() -> impl Parser<easy::Stream<&'t str>, Output = Expr> {
+    spaces().with((
+        function_name().skip(spaces()),
+        between(
+            char('(').skip(spaces()),
+            char(')').skip(spaces()),
+            sep_by(rib_expr().skip(spaces()), char(',').skip(spaces())),
+        ),
+    )
+        .map(|(name, args)| Expr::Call(name, args))
+        .message("Unable to parse call"))
+}
+
+// TODO;  Reusing function_name between Rib and internals of GOLEM may be a surface level requirement
+// as users can form function name in various other ways.
+// Example: Arguments to a resource can be partial because they may come from request parameters
+// and these are not represented using the current structure of ParsedFunctionName
+pub fn function_name<'t>() -> impl Parser<easy::Stream<&'t str>, Output = ParsedFunctionName> {
+    let identifier = || many1(alpha_num().or(token('-'))).map(|string: String| string);
+    let namespace = many1(identifier()).message("namespace");
+    let package = many1(identifier()).message("package");
+    let ns_pkg = (namespace, token(':'), package).map(|(ns, _, pkg)| (ns, pkg));
+    let interface = many1(identifier()).message("interface");
+
+    let capture_resource_params = || {
+        parser(|input| {
+            let _: &mut easy::Stream<&str> = input;
+            let mut nesting = 1;
+            let mut current_param = String::new();
+            let mut result = Vec::new();
+            let mut result_committed: Option<Commit<()>> = None;
+            while nesting > 0 {
```

Comment:
I think this termination condition is wrong as matching brace may be missing forever and this never ends. I will see to it.